src/node/rpc/test/proposal_id_test.cpp

```diff
@@ -0,0 +1,333 @@
+}
+
+auto init_frontend(
+  NetworkState& network,
+  GenesisGenerator& gen,
+  StubNodeState& node,
+  ShareManager& share_manager,
+  const int n_members,
+  std::vector<tls::Pem>& member_certs)
+{
+  // create members
+  for (uint8_t i = 0; i < n_members; i++)
+  {
+    member_certs.push_back(get_cert(i, kp));
+    gen.activate_member(gen.add_member(member_certs.back()));
+  }
+
+  set_whitelists(gen);
+  gen.set_gov_scripts(lua::Interpreter().invoke<json>(gov_script_file));
+  gen.finalize();
+
+  return MemberRpcFrontend(network, node, share_manager);
+}
+
+DOCTEST_TEST_CASE("Proposer ballot")
+{
+  NetworkState network;
+  network.tables->set_encryptor(encryptor);
+  auto history = std::make_shared<ccf::NullTxHistory>(*network.tables, 0, *kp);
+  network.tables->set_history(history);
+  auto consensus = std::make_shared<aft::StubCFTConsensus>();
+  network.tables->set_consensus(consensus);
+  auto gen_tx = network.tables->create_tx();
+  GenesisGenerator gen(network, gen_tx);
+  gen.init_values();
+  gen.create_service({});
+
+  const auto proposer_cert = get_cert(0, kp);
+  const auto proposer_id = gen.add_member(proposer_cert);
+  gen.activate_member(proposer_id);
+  const auto voter_cert = get_cert(1, kp);
+  const auto voter_id = gen.add_member(voter_cert);
+  gen.activate_member(voter_id);
+
+  set_whitelists(gen);
+  gen.set_gov_scripts(lua::Interpreter().invoke<json>(gov_script_file));
+  gen.finalize();
+
+  ShareManager share_manager(network);
+  StubNodeState node;
+  MemberRpcFrontend frontend(network, node, share_manager);
+
+  frontend.open();
+
+  {
+    DOCTEST_INFO("Propose, no votes");
+    const auto proposed_member = get_cert(2, kp);
+
+    Propose::In proposal;
+    proposal.script = std::string(R"xxx(
+      tables, member_info = ...
+      for i = 1,10000000,1
+      do
+      u = i ^ 0.5
+      end
+      return Calls:call("new_member", member_info)
+    )xxx");
+    proposal.parameter["cert"] = proposed_member;
+    proposal.parameter["encryption_pub_key"] = dummy_enc_pubk;
+    const auto propose =
+      create_signed_request(proposal, "proposals", kp, proposer_cert);
+
+    Propose::Out out1;
+    Propose::Out out2;
+
+    auto fn = [](
+                MemberRpcFrontend& f,
+                const std::vector<uint8_t>& r,
+                const tls::Pem& i,
+                Propose::Out& o) {
+      const auto rs = frontend_process(f, r, i);
+      o = parse_response_body<Propose::Out>(rs);
+    };
+
+    auto t1 = std::thread(
+      fn,
+      std::ref(frontend),
+      std::ref(propose),
+      std::ref(proposer_cert),
+      std::ref(out1));
+    auto t2 = std::thread(
+      fn,
+      std::ref(frontend),
+      std::ref(propose),
+      std::ref(proposer_cert),
+      std::ref(out2));
+    t1.join();
+    t2.join();
+
+    DOCTEST_CHECK(out1.state == ProposalState::OPEN);
+    DOCTEST_CHECK(out2.state == ProposalState::OPEN);
+    DOCTEST_CHECK(out1.proposal_id != out2.proposal_id);
```

Comment:
This needs an extra check to confirm we have triggered a conflict, perhaps looking at a (yet to be added) conflict counter in the endpoint metrics.